File: adaptive_repair/data/python_programs/wrap.py
def wrap(text, cols):
    lines = []
    while len(text) > cols:
        # Search for the last space within the current column width (inclusive of cols index)
        end = text.rfind(' ', 0, cols + 1)

        if end == -1: # No space found in the first 'cols' characters
            # This means a word is longer than 'cols', so we must break the word.
            end = cols
        elif end == 0 and len(text) > 1: # Only space found is at the beginning, and text is not just a single space
            # This means the text starts with a space, and the subsequent word might exceed 'cols'.
            # To ensure progress and consume the leading space, we take just the space.
            end = 1
        # else: end is a valid space index > 0, which is fine.

        line, text = text[:end], text[end:]
        lines.append(line)

    # After the loop, append any remaining text. This segment will have a length
    # less than or equal to 'cols', making it a valid final line.
    if text:
        lines.append(text)

    return lines

File: adaptive_repair/data/java_programs/WRAP.java
package java_programs;

import java.util.ArrayList;
import java.util.List;

public class WRAP {

    public static void main(String[] args) {
        // This was a test line in the original Python code to understand rfind behavior.
        // The equivalent in Java for "abc".rfind("c", 0, 30 + 1) is "abc".lastIndexOf("c", 30).
        System.out.println("abc".lastIndexOf("c", 30));
    }

    public static List<String> wrap(String text, int cols) {
        List<String> lines = new ArrayList<>();

        while (text.length() > 0) { // Loop as long as there's text remaining
            if (text.length() <= cols) {
                // The remaining text fits on one line, so add it and we're done.
                lines.add(text);
                text = ""; // Mark text as fully processed
                break; // Exit the loop
            }
            
            // If text is longer than cols, try to find a break point
            // Java's lastIndexOf(search_string, from_index) searches backward from from_index.
            // This means it searches within text[0...from_index].
            // This matches Python's rfind(sub, 0, cols + 1) which searches text[0...cols].
            int end = text.lastIndexOf(" ", cols);

            if (end == -1) {
                // No space found within the first 'cols' characters, so break the word.
                end = cols;
            }
            
            String line = text.substring(0, end);
            text = text.substring(end);
            
            // After splitting, remove any leading spaces from the remaining text
            // to ensure the next line doesn't start with a space and to prevent
            // infinite loops if 'end' pointed to a space at the beginning of 'text'.
            text = text.stripLeading(); // Equivalent to Python's lstrip()
            
            lines.add(line);
        }
        
        return lines;
    }
}